lib/solver.cpp

```diff
@@ -368,108 +370,125 @@ namespace quda {
     }
   }
 
-  double Solver::stopping(double tol, double b2, QudaResidualType residual_type)
+  vector<double> Solver::stopping(double tol, cvector<double> &b2, QudaResidualType residual_type)
   {
-    double stop=0.0;
+    vector<double> stop(b2.size(), 0.0);
     if ( (residual_type & QUDA_L2_ABSOLUTE_RESIDUAL) &&
 	 (residual_type & QUDA_L2_RELATIVE_RESIDUAL) ) {
-      // use the most stringent stopping condition
-      double lowest = (b2 < 1.0) ? b2 : 1.0;
-      stop = lowest*tol*tol;
+      for (auto i = 0u; i < b2.size(); i++) {
+        // use the most stringent stopping condition
+        double lowest = (b2[i] < 1.0) ? b2[i] : 1.0;
+        stop[i] = lowest * tol * tol;
+      }
     } else if (residual_type & QUDA_L2_ABSOLUTE_RESIDUAL) {
-      stop = tol*tol;
+      for (auto i = 0u; i < b2.size(); i++) stop[i] = tol * tol;
+    } else if (residual_type & QUDA_L2_RELATIVE_RESIDUAL) {
+      for (auto i = 0u; i < b2.size(); i++) stop[i] = b2[i] * tol * tol;
     } else {
-      stop = b2*tol*tol;
+      // if invalid residual then convergence is set by iteration count only
+      for (auto i = 0u; i < b2.size(); i++) stop[i] = 0.0;
     }
 
     return stop;
   }
 
-  bool Solver::convergence(double r2, double hq2, double r2_tol, double hq_tol) {
-
-    // check the heavy quark residual norm if necessary
-    if (param.residual_type & QUDA_HEAVY_QUARK_RESIDUAL) {
-      if (std::isnan(hq2) || std::isinf(hq2))
-        errorQuda("Solver appears to have diverged with heavy quark residual %9.6e", hq2);
-
-      if (hq2 > hq_tol) return false;
-    }
-
-    // check the L2 relative residual norm if necessary
-    if ((param.residual_type & QUDA_L2_RELATIVE_RESIDUAL) || (param.residual_type & QUDA_L2_ABSOLUTE_RESIDUAL)) {
-      if (std::isnan(r2) || std::isinf(r2)) errorQuda("Solver appears to have diverged with residual %9.6e", r2);
+  bool Solver::convergence(cvector<double> &r2, cvector<double> &hq2, cvector<double> &r2_tol, cvector<double> &hq_tol)
+  {
+    for (auto i = 0u; i < r2.size(); i++) {
+      // check the heavy quark residual norm if necessary
+      if (param.residual_type & QUDA_HEAVY_QUARK_RESIDUAL) {
+        if (std::isnan(hq2[i]) || std::isinf(hq2[i]))
+          errorQuda("Solver appears to have diverged with heavy quark residual %9.6e", hq2[i]);
+        if (hq2[i] > hq_tol[i]) return false;
+      }
 
-      if (r2 > r2_tol) return false;
+      // check the L2 relative residual norm if necessary
+      if ((param.residual_type & QUDA_L2_RELATIVE_RESIDUAL) || (param.residual_type & QUDA_L2_ABSOLUTE_RESIDUAL)) {
+        if (std::isnan(r2[i]) || std::isinf(r2[i])) errorQuda("Solver appears to have diverged with residual %9.6e", r2[i]);
+        if (r2[i] > r2_tol[i]) return false;
+      }
     }
-
     return true;
   }
 
-  bool Solver::convergenceHQ(double, double hq2, double, double hq_tol)
+  bool Solver::convergenceHQ(cvector<double> &hq2, cvector<double> &hq_tol)
```

Comment:
There's a big "gotcha" that happens here: in, for example, `inv_cg_quda.cpp`, this routine is called as:

```
convergenceHQ(hq_res, param.tol_hq)
```

Where `param.tol_hq` is just a _double_. Within `convergenceHQ` (and elsewhere), this is blindly converted to a `cvector<double>` where the 0th element is `param.tol_hq` and then the rest are uninitialized. This gives garbage for the other values, messing up convergence.

I think the only way to catch this is to make a conversion from a type `T` to `cvector<T>` require an explicit cast, which I think you've imposed on other types as well.

Reply:
50e78799c7b9e5babb2ced06bcf2f27e35f036e3 addresses this issue: the convergence and printing helper functions now check the sizes of the arguments and will call a sane error if the sizes don't match.